Modules/sha256module.c

```diff
@@ -681,43 +681,44 @@ static struct PyMethodDef SHA_functions[] = {
     {NULL,      NULL}            /* Sentinel */
 };
 
-
-/* Initialize this module. */
-
-static struct PyModuleDef _sha256module = {
-        PyModuleDef_HEAD_INIT,
-        "_sha256",
-        NULL,
-        -1,
-        SHA_functions,
-        NULL,
-        NULL,
-        NULL,
-        NULL
-};
-
-PyMODINIT_FUNC
-PyInit__sha256(void)
+static int sha256_exec(PyObject *module)
 {
-    PyObject *m;
-
     Py_SET_TYPE(&SHA224type, &PyType_Type);
     if (PyType_Ready(&SHA224type) < 0) {
-        return NULL;
+        return -1;
     }
     Py_SET_TYPE(&SHA256type, &PyType_Type);
     if (PyType_Ready(&SHA256type) < 0) {
-        return NULL;
+        return -1;
     }
 
-    m = PyModule_Create(&_sha256module);
-    if (m == NULL)
-        return NULL;
-
     Py_INCREF((PyObject *)&SHA224type);
-    PyModule_AddObject(m, "SHA224Type", (PyObject *)&SHA224type);
+    PyModule_AddObject(module, "SHA224Type", (PyObject *)&SHA224type);
     Py_INCREF((PyObject *)&SHA256type);
-    PyModule_AddObject(m, "SHA256Type", (PyObject *)&SHA256type);
-    return m;
+    PyModule_AddObject(module, "SHA256Type", (PyObject *)&SHA256type);
```

Comment:
MAYBE you can move those typeobject to module state(you can use a separate PR to do this job).

something like this PR: https://github.com/python/cpython/commit/ec689187957cc80af56b9a63251bbc295bafd781

Reply:
It's ok to convert static types to heap types in a separated PR.

Reply:
@shihai1991 What is the advantage of that?  Does this help with subintepreter issues?

Reply:
@shihai1991 oh, I think I get it.  static ptrs are global state that can not be isolated to a given subinterpreter.  I can give it a shot but maybe in another PR if I can't figure out how right away